### PR TITLE
Disallow `new MyObj.type`

### LINKED ISSUE #12322: object generation compiles successfully.
## reproduction steps

using Scala 2.13.4,

I can compile by instantiating an object.
The following code produces a `java.lang.IllegalAccessError` at runtime:

```scala
// ObjectNew.scala

object MyObj {
  val a: Int = 123
  val b: Double = 456.789
  val c: String = "ABC"
}

object ObjectNew {
  def main(args: Array[String]): Unit = {
    val mo: MyObj.type = new MyObj.type() // here
  }
}
```

```shell-session
# scalac -version
Scala compiler version 2.13.4 -- Copyright 2002-2020, LAMP/EPFL and Lightbend, Inc.
# scalac ObjectNew.scala
# scala ObjectNew
Exception in thread "main" java.lang.IllegalAccessError: tried to access method MyObj$.<init>()V from class Main$
        at Main$.main(ObjectNew.scala:11)
        at Main.main(ObjectNew.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at scala.reflect.internal.util.ScalaClassLoader.$anonfun$run$2(ScalaClassLoader.scala:105)
        at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:40)
        at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:37)
        at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:130)
        at scala.reflect.internal.util.ScalaClassLoader.run(ScalaClassLoader.scala:105)
        at scala.reflect.internal.util.ScalaClassLoader.run$(ScalaClassLoader.scala:97)
        at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:130)
        at scala.tools.nsc.CommonRunner.run(ObjectRunner.scala:29)
        at scala.tools.nsc.CommonRunner.run$(ObjectRunner.scala:28)
        at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:43)
        at scala.tools.nsc.CommonRunner.runAndCatch(ObjectRunner.scala:35)
        at scala.tools.nsc.CommonRunner.runAndCatch$(ObjectRunner.scala:34)
        at scala.tools.nsc.AbstractScriptRunner.runCompiled(ScriptRunner.scala:168)
        at scala.tools.nsc.AbstractScriptRunner.$anonfun$runScript$1(ScriptRunner.scala:177)
        at scala.tools.nsc.AbstractScriptRunner.$anonfun$withCompiledScript$9(ScriptRunner.scala:154)
        at scala.tools.nsc.util.package$.waitingForThreads(package.scala:55)
        at scala.tools.nsc.AbstractScriptRunner.runScript(ScriptRunner.scala:151)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:74)
        at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:91)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:103)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:108)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```

In Scala 2.13.3 or older, a compile error.

```shell-session
[root@localhost ~]# scalac -version
Scala compiler version 2.13.3 -- Copyright 2002-2020, LAMP/EPFL and Lightbend, Inc.
[root@localhost ~]# scalac ObjectNew.scala
ObjectNew.scala:11: error: class type required but MyObj.type found
    val mo: MyObj.type = new MyObj.type()
                                  ^
1 error
```

## problem

Unable to detect object instantiation error at compile time.

### PR DESCRIPTION
**Pull request by Jasper-M**
... while still allowing `classOf[MyObj.type]`.

Fixes scala/bug#12322

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -370,9 +370,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     private def errorNotStable(tpt: Tree, found: Type) = { TypeNotAStablePrefixError(tpt, found); false }
 
     /** Check that `tpt` refers to a non-refinement class type */
-    def checkClassType(tpt: Tree): Boolean = {
+    def checkClassOrModuleType(tpt: Tree): Boolean = {
       val tpe = unwrapToClass(tpt.tpe)
-      isNonRefinementClassType(tpe) || errorNotClass(tpt, tpe)
+      isNonRefinementClassType(tpe, allowModules = true) || errorNotClass(tpt, tpe)
```
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Does everything pass (the new test and the `classOf` test) by subsuming the `case SingleType(_, sym) if sym.isModule => true` additional case here?  Seems cleaner and more local than "generalising" a boolean function with a new boolean parameter.

**Review by dwijnand — APPROVED** [review]
LGTM, thanks!

File: src/reflect/scala/reflect/internal/Types.scala [hunk 2]
```diff
@@ -4763,7 +4763,7 @@ trait Types
   /** def isNonValueType(tp: Type) = !isValueElseNonValue(tp) */
 
   def isNonRefinementClassType(tpe: Type) = tpe match {
-    case SingleType(_, sym) => sym.isModuleOrModuleClass
+    case SingleType(_, sym) => sym.isModuleClass
```
Comment (dwijnand) [inline src/reflect/scala/reflect/internal/Types.scala:4766]: For the record: this reverts the change here that came from #9279, by moving it locally to `typedClassOf`.

**Comment by dwijnand** [comment]
This is just a bug fix.  #9279 was the improvement which we shipped and release noted in 2.13.4.

**Comment by dwijnand** [comment]
"Just" in the "no need to release note it" sense 🙇🏼 thank you for the fix!!

**Comment by Jasper-M** [comment]
💔

**Pull request merged**